src/main/java/xbot/common/subsystems/vision/AprilTagVisionSubsystem.java

```diff
@@ -0,0 +1,228 @@
+// Copyright 2021-2025 FRC 6328
+// http://github.com/Mechanical-Advantage
+//
+// This program is free software; you can redistribute it and/or
+// modify it under the terms of the GNU General Public License
+// version 3 as published by the Free Software Foundation or
+// available in the root directory of this project.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+package xbot.common.subsystems.vision;
+
+import edu.wpi.first.apriltag.AprilTagFieldLayout;
+import edu.wpi.first.math.Matrix;
+import edu.wpi.first.math.VecBuilder;
+import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.math.geometry.Pose3d;
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.numbers.N1;
+import edu.wpi.first.math.numbers.N3;
+import edu.wpi.first.wpilibj.Alert;
+import edu.wpi.first.wpilibj.Alert.AlertType;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import org.littletonrobotics.junction.Logger;
+import xbot.common.injection.electrical_contract.XCameraElectricalContract;
+import xbot.common.properties.DoubleProperty;
+import xbot.common.properties.PropertyFactory;
+
+import java.util.LinkedList;
+import java.util.List;
+
+public class AprilTagVisionSubsystem extends SubsystemBase {
+    private final VisionConsumer consumer;
+    private final AprilTagFieldLayout aprilTagFieldLayout;
+    private final AprilTagVisionIO[] io;
+    private final VisionIOInputsAutoLogged[] inputs;
+    private final Alert[] disconnectedAlerts;
+
+    // Basic filtering thresholds
+    private final DoubleProperty maxAmbiguity;
+    private final DoubleProperty maxZError;
+
+    // Standard deviation baselines, for 1 meter distance and 1 tag
+    // (Adjusted automatically based on distance and # of tags)
+    private final DoubleProperty linearStdDevBaseline;
+    private final DoubleProperty angularStdDevBaseline;
+
+    // Multipliers to apply for MegaTag 2 observations
+    private final DoubleProperty linearStdDevMegatag2Factor;
+    private final DoubleProperty angularStdDevMegatag2Factor;
+
+    // Standard deviation multipliers for each camera
+    // (Adjust to trust some cameras more than others)
+    public static double[] cameraStdDevFactors =
+            new double[] {
+                    1.0, // Camera 0
+                    1.0 // Camera 1
+            };
+
+    public AprilTagVisionSubsystem(VisionConsumer consumer, PropertyFactory pf, AprilTagFieldLayout fieldLayout, XCameraElectricalContract contract) {
```

Comment:
❓  Was it intentional that these not be injectable? Would the idea be to build it by hand in the Robot class?

Reply:
I forgot to add it

Reply:
⭐ ⭐ ⭐ fieldLayout and VisionConsumer need to be added to the dagger components for this to be injectable right?